pallas-rolldb/src/wal/stream.rs

```diff
@@ -1,31 +1,49 @@
 use futures_core::Stream;
 
-use super::{BlockHash, BlockSlot, Log, Store};
+use crate::kvtable::Error;
+
+use super::{Log, Store};
 
 pub struct RollStream;
 
 impl RollStream {
-    pub fn intersect(
+    pub fn stream_wal(
         store: Store,
-        intersect: Vec<(BlockSlot, BlockHash)>,
-    ) -> impl Stream<Item = Log> {
-        async_stream::stream! {
-            let mut last_seq = None;
+        last_wal_seq: Option<u64>,
+    ) -> impl Stream<Item = Result<Log, Error>> {
+        async_stream::try_stream! {
+            let mut last_seq = last_wal_seq;
+
+            let iter = store.crawl_after(last_wal_seq);
+
+            for entry in iter {
+                let (wal_seq, log) = entry?;
 
-            let iter = store.crawl_from_intersect(&intersect).unwrap();
+                if let Some(prev_seq) = last_seq {
+                    if wal_seq != (prev_seq + 1) {
+                        Err(Error::UnexpectedWalSeq(prev_seq + 1, wal_seq))?
+                    }
+                };
```

Comment:
why do we need this check?

Reply:
This is what protects against the race condition where the found seq for the intersect has been pruned from the WAL

Reply:
Sure, for that exact scenario we would only need to check the first entry returned by the iterator, but I figured why not check all of them